File: solutions/typescript/bank-account/1/bank-account.ts
export class ValueError extends Error {
  constructor(message?: string) {
    super(message || 'Bank account error')
  }
}

export class BankAccount {
  private isOpen: boolean = false
  #balance: number = 0
  
  constructor() {}

  open(): void | never {
    if (this.isOpen) {
      throw new ValueError('Account is already open')
    }
    this.isOpen = true
  }

  close(): void {
    if (!this.isOpen) {
      throw new ValueError('Account is already closed')
    }
    this.isOpen = false
    this.#balance = 0
  }

  deposit(amount: number): void | never {
    if (!this.isOpen) {
      throw new ValueError('Account is closed')
    }
    if (amount < 0) {
      throw new ValueError('Cannot deposit negative amount')
    }
    this.#balance += amount
  }

  withdraw(amount: number): void | never {
    if (!this.isOpen) {
      throw new ValueError('Account is closed')
    }
    if (amount < 0) {
      throw new ValueError('Cannot withdraw negative amount')
    }
    if (this.#balance < amount) {
      throw new ValueError('Insufficient balance to withdraw from')
    }
    this.#balance -= amount
  }

  get balance(): number {
    if (!this.isOpen) {
      throw new ValueError('Account is closed')
    }
    return this.#balance
  }
}

File: solutions/python/bank-account/1/bank_account.py
class BankAccount:
    def __init__(self):
        self._open = False
        self._balance = 0

    def get_balance(self):
        if not self._open:
            raise ValueError("account not open")
        return self._balance

    def open(self):
        if self._open:
            raise ValueError("account already open")
        self._open = True

    def deposit(self, amount):
        if not self._open:
            raise ValueError("account not open")
        if amount < 0:
            raise ValueError("amount must be greater than 0")
        self._balance += amount

    def withdraw(self, amount):
        if not self._open:
            raise ValueError("account not open")
        if amount < 0:
            raise ValueError("amount must be greater than 0")
        if self._balance < amount:
            raise ValueError("amount must be less than balance")
        self._balance -= amount

    def close(self):
        if not self._open:
            raise ValueError("account not open")
        self._open = False
        self._balance = 0